services/service.ts

```
@@ -0,0 +1,355 @@
+  telnetDisabled: 'webosbrew_telnet_disabled',
+  failsafe: 'webosbrew_failsafe',
+  sshdEnabled: 'webosbrew_sshd_enabled',
+  blockUpdates: 'webosbrew_block_updates',
+} as Record<string, FlagName>;
+
+function asyncCall<T extends Record<string, any>>(srv: Service, uri: string, args: Record<string, any>): Promise<T> {
+  return new Promise((resolve, reject) => {
+    srv.call(uri, args, ({ payload }) => {
+      if (payload.returnValue) {
+        resolve(payload as T);
+      } else {
+        reject(payload);
+      }
+    });
+  });
+}
+
+function createToast(message: string): Promise<Record<string, any>> {
+  return asyncCall(service, 'luna://com.webos.notification/createToast', {
+    sourceId: serviceInfo.id,
+    message,
+  });
+}
+
+/**
+ * Generates local file checksum.
+ */
+async function hashFile(filePath: string, algorithm: string): Promise<string> {
+  const download = fs.createReadStream(filePath);
+  const hash = createHash(algorithm, { encoding: 'hex' });
+  await asyncPipeline(download, hash);
+  return hash.read();
+}
+
+/**
+ * Elevates a package by name.
+ */
+async function elevateService(pkg: string) {
+  if (runningAsRoot()) {
+    console.info('Elevating service...');
+    await asyncExecFile(path.join(__dirname, 'elevate-service'), [pkg]);
+  } else {
+    console.error('Trying to elevate service without running as root. Skipping.');
+  }
+}
+
+/**
+ * Returns the file path for a flag.
+ */
+function flagPath(flag: FlagName): string {
+  return `/var/luna/preferences/${flag}`;
+}
+
+/**
+ * Returns whether a flag is set or not.
+ */
+async function flagRead(flag: FlagName): Promise<boolean> {
+  const err = await asyncAccess(flagPath(flag));
+  return !err;
+}
+
+/**
+ * Sets the value of a flag.
+ */
+async function flagSet(flag: FlagName, enabled: boolean) {
+  if (enabled) {
+    // Value is ignored, file presence is what matters. Using 1 as a hint.
+    await asyncWriteFile(flagPath(flag), '1');
+  } else {
+    // Error is ignored.
```

...is it? That sounds... weird? Do you have any idea where this behaviour comes from? Plain `fs.unlink` returns an error if the file doesn't exist, and thus the promisified version should throw as well. Or is this some weird behaviour of `Promise.all` ignoring the errors?